old/disksetup/disksetup.8
.\" Copyright (c) 1993,1994 Berkeley Software Design, Inc.
.\" All rights reserved.
.\"
.\" BSDI	disksetup.8,v 2.1 1995/02/03 06:40:01 polk Exp
.\"
.TH DISKSETUP 8 "February 18,1993"
.UC 8
.SH NAME
disksetup \- read and write disk labels
.SH SYNOPSIS
.B disksetup
.I disk
.br
.B disksetup \-e 
[
.B \-I
[\c
.BR \-I ]
]
.I disk
[
.I xxboot
.I bootxx
[
.I mboot
] ]
.br
.B disksetup \-i
[
.B \-I
[\c
.BR \-I ]
]
.I disk
.br
.B disksetup \-R
.I disk
.I protofile
[
.I xxboot
.I bootxx
[
.I mboot
] ]
.br
.B disksetup 
[
.B \-NW
]
.I disk
.br
.B disksetup \-B
.I disk
[
.I xxboot
.I bootxx
[
.I mboot
] ]
.SH DESCRIPTION
.PP
The goal of the
.I disksetup
program is to define the FDISK table if needed (see below), the BSD
partition tables (and geometry information), and to write the appropriate
boot blocks to the disk.
.PP
.I Disksetup
installs, examines, or modifies both the FDISK and BSD/OS disk labels on
a disk drive.
.I Disksetup
can set or change the disk geometry, the sizes, types and offsets for the
FDISK and BSD/OS partitions.  It can also write both BSD/OS boot blocks
and, optionally, master boot record boot blocks.
.PP
A disk shared between BSD/OS and any other operating system(s) has a
single FDISK partition table in its first physical sector.  This table
defines 1 to 4 partitions which enumerate the extents of the sections of
the disk used by various operating systems and the types of those
sections.  Additionally, the FDISK table indicates which partition is
currently `active' and will be used for the next reboot (unless
overridden).  The FDISK partition table is used by BSD/OS only to find
the location of the BSD part of the disk.  The four FDISK partitions are
numbered `1' through `4'.
.PP
Disks used by BSD/OS also have a BSD label (in sector 1 of the bootable
BSD partition) which describes the geometry of the entire disk (number of
heads, sectors, tracks, etc.) and enumerates 1 to 8 BSD partitions which
are named `a' through `h'.
.PP
BSD/OS uses the BSD partition definitions to access the disk (including
access of MS-DOS files in other FDISK partitions).
.PP
In common usage, BSD/OS divides its part of the primary disk into three
sections:  one for the root filesystem (always the `a' partition), one for
swapping (and paging) (always the `b' partition), and one for /usr and
other files (usually the `h' partition).  If BSD/OS is to access other
parts of the disk (e.g., for MS-DOS co-residency), another partition is
assigned to refer to that disk data (often the `d' partition).
.PP
More advanced users sometimes set up additional partitions, e.g., for
the /var directories.
.PP
Confusingly enough, BSD/OS always defines one more partition:  the `c'
partition.  The `c' partition refers to the entire disk (including the
boot blocks and the bad sector table at the end of the disk).  This
partition obviously overlaps with any other BSD partition on the disk.
The `c' partition can be used to define a filesystem only on non-primary
disks and only on disks that have no bad-block table (e.g., SCSI).  The
`c' partition must not be used to store BSD/OS filesystems on IDE or ESDI
drives (drives using the 
.I wd 
disk driver).  Instead, use a different,
slightly shorter partition (e.g., define `h' to be the entire usable
portion of the disk).
In the interactive (\c
.BR \-i )
mode,
.I disksetup
enforces the correct maximum size.
.SH "INVOKING \fIdisksetup\fP IN THE INTERACTIVE MODE
.PP
.I Disksetup
is most often used in the interactive mode.  Invoke it by specifying
the
.B \-i
flag and naming the disk to be set up.  The example below shows
invocations for SCSI and non-SCSI primary drives, respectively:
.br
.nf
.ft CW
	# disksetup -i sd0
\fPor\fP
	# disksetup -i wd0
.br
.fi
.ft
.sp 0.5
.PP
If the 
.B \-I
option is also specified, 
.I disksetup
will ignore any existing BSD label on the disk.  This can be very useful
if an incorrect geometry was previously specified, or if the current
label is valid but incorrect for some reason.
.PP
If two
.B \-I
arguments are specified, 
.I disksetup
will ignore not only any existing BSD label, but any existing FDISK
label as well.  It will also open the device with the O_NONBLOCK
argument to inhibit the kernel from reading any existing label.
.PP
Other options are described below.
.PP
.I disksetup
asks if the disk is to be used exclusively for BSD partitions or if it is
to be shared with other operating systems.  Answer the question
appropriately.
.PP
.I disksetup
learns the drive's geometry either by reading it from an existing BSD
label on the drive or by querying the user.  If no BSD label previously
existed, no geometry information is available so
.I disksetup
asks the user to supply one.  Otherwise,
.I disksetup
proceeds to the sizing screens (see below).
.PP
When no geometry information is available,
.I disksetup
asks the user to choose from:
.in +\w'\(bu\ 'u
.ti -\w'\(bu\ 'u
\(bu\ running the
.I scsicmd
utility (only for SCSI disks, of course) to query the drive for its
geometry,
.ti -\w'\(bu\ 'u
\(bu\ supplying the name of a `disktab' entry that corresponds to the disk
to be labeled (chosen from the file /etc/disktab),
.ti -\w'\(bu\ 'u
\(bu\ querying the kernel for the current geometry (the kernel may have
information from an on-disk label or from the CMOS),
.ti -\w'\(bu\ 'u
\(bu\ querying an IDE/ESDI controller for the geometry,
.ti -\w'\(bu\ 'u
\(bu\ answering prompts about the geometry,
.ti -\w'\(bu\ 'u
\(bu\ naming a file containing a saved geometry (from a previous
invocation of disksetup (described later) or some other command)
.in -\w'\(bu\ 'u
.PP
Except for supplying the answers to prompts, all choices above proceed
`automatically'.  When answering the prompts, the user must supply the
following data:  interface type, vendor and type (e.g., `Maxtor P012S')
which is only used to document the disk's vendor and type ID, bytes/sector
(almost always 512, always a multiple of 512), number of cylinders (from the
manual; must be the number of data cylinders on the drive \- it excludes
those cylinders for the drive's own internal use such as for bad blocks),
number of heads (also known as tracks/cylinder; from the manual), number
of sectors per track (from the manual), and total data sectors (which
might not be the same as the default in the unlikely case that the last
cylinder is not a complete one).  The user then verifies that the
information is correct before proceeding.
.SH "INTERACTIVE SIZING
.PP
.I Disksetup
displays one of two different screen sets for sizing, depending on
whether the drive is to be shared between operating systems.  If the drive
is used exclusively for BSD/OS, then
.I disksetup
displays the `BSD Partition Sizing' screen.  Otherwise, it displays the
`FDISK Partition Sizing & BSD Partition Assignment' screen.
Separate sections that follow document these two different execution
paths.
.PP
In general, the interactive screens use `hot keys' for their commands
and otherwise require a `return' to be entered after the input.
A `return' with no input usually aborts the current operation;
otherwise, it indicates acceptance of a default value which has been
presented between square brackets.  Your backspace and line delete
characters work correctly for input to these screens.
.PP
Numbers can be entered in many ways:  absolute, relative, and with a unit
suffix.  Absolute numbers are entered without a sign and have a value
identical to the number typed.  Relative numbers are preceded by a + or \-
sign and specify a change to the current value (i.e., an increase or
decrease). 
.PP
Normally, numbers represent sectors,  However, specifying a unit suffix
can change the meaning of the numbered typed.  Use the `m' suffix to
specify megabytes (1024*1024 bytes); `c' for cylinders, or `M' for
cylinder-aligned-megabytes (the entered number multiplied by 1024*1024 and
rounded up to the nearest cylinder boundary.
.br
.ne 23
.SH "BSD PARTITION SIZING SCREEN'
.PP
In the BSD-only scenario, the sizing screen resembles this:
.ft CW
.nf
                       BSD Partition Sizing
.sp 0.5
BSD Type      Size    MB     Start     End  Warnings
 a  4.2      33600 (  16)        0   33599
 b  swap    134400 (  65)    33600  167999
 c  Unused  639450 ( 312)        0  639449
 d  4.2      67200 (  32)   168000  235199
 e  4.2      67200 (  32)   235200  302399
 f   <unused>
 g   <unused>
 h  4.2     337050 ( 164)   302400  639449
                Last Data Sector:   639449
.sp 0.5
Commands:  [T]ype  [S]tart  [L]ength  [N]ext phase
Command>
.sp 0.5
Directions:
   Use `L' and `S' to set the BSD partition Lengths and Starting sectors
   Use `T' to set the BSD partition Types
   Use `N' to move to the next phase:  Installing boot blocks
   Use `X' to abort disksetup
Numbers can be relative (with + or \-) or suffixed with c (cylinders),
     m (megabytes), or M (megabytes rounded to cylinders)
.fi
.ft
.sp 0.5
The goal of this screen is to set the sizes, starting offsets (i.e.,
starting sector numbers), and types of the BSD partitions.  All sizes
and offsets are in terms of sectors.
.PP
This screen allows you to set up overlapping partitions and will
display a warning in the `Warnings' column.  Unless you have specific
needs for overlapping partitions, they are best avoided.
.PP
Use the `T', `S', and `L' keys to set the parameters you desire.  As a
starting idea, consider using an 8 MB `a' (root) partition, 16 MB `b'
(swap) partition (32MB if you will be using the X window system or MFS,
the memory based filesystem), and allocating the remainder to the `h'
partition for /usr directory and files owned by your system's users.
.PP
Enter `N' when the parameters shown meet your needs to move to the
next phase:  boot block installation (if you are using BSD partitions
only; skip the next section and move on to the discussion of boot
blocks).
.br
.ne 26
.SH "FDISK PARTITION SIZING & BSD PARTITION ASSIGNMENT SCREENS
.PP
The first FDISK & BSD sizing screen resembles this \- though your system
may already have FDISK and/or BSD partitions which would then appear
in the table (start and length values are in terms of sectors):
.ft CW
.nf
            FDISK Partition Sizing & BSD Partition Assignment
FDISK
Part#  type        Start Length    MB BSD Partitions    Warnings
  1  [--unused--]
  2  [--unused--]
  3  [--unused--]
  4  [--unused--]
Last data sector: 414364                               Gap at end: 414365
.sp 0.5
Unassigned BSD partitions:  a b d e f g h
.sp 0.5
Commands:  [T]ype  [S]tart  [L]ength  [A]ssign BSD partition  [N]ext
phase
.sp 0.5
Command>
.sp 0.5
Directions:
   Use `L' and `S' to set the FDISK partition Lengths and Starting sectors
   Use `T' to set the FDISK partition Types
   Use `A' to assign the BSD partitions to cover the FDISK partitions
   -- One BSD partition for non-BSDI FDISK partitions -- One or more for BSDI
   -- Assignment is to the end of the partition list
   Use `N' to move to the next phase:  BSD partition sizing
   Use `X' to abort disksetup
.ft
.sp 0.5
.br
.fi
This screen has two goals.  It sets the starting sector, length, and type
for each FDISK partition to be used.  It also assigns the name of the
BSD partition(s) to be associated with each FDISK partition.
.PP
The \fIdisksetup\fP program does \fBnot\fP allow arbitrary disk layouts;
instead it enforces common sense rules.  For instance, \fIdisksetup\fP
does not allow partitions to overlap in any final configuration.
.PP
Automatically enforced rules:  \fIDisksetup\fP occasionally adjusts start
addresses to various boundaries (a warning is shown).  Only the BSDI parts
of the FDISK partition can have more than one associated BSD partition.
.PP
Figure out your disk's overall layout scheme (how much for MS/DOS; how
much for BSD/OS; how much for any other system).  Enter the parameters
using the `L' key to enter the FDISK partition size (do this first).
Next, set the FDISK partition types using the `T' key.  Then, use `S' to
enter the starting sectors.  Adhering to this order will minimize
distractions when partition starting sectors are moved due to constraints
inherent in the various operating systems.
.PP
Adjust the lengths and starting sectors to eliminate all warnings.
.PP
Finally, assign the BSD partition names (`a' through `h') that you wish to
use to refer to the various partitions (and sub-partitions in the case of
a BSD partition).  A typical scheme for BSD partition layout is:  `a',
`b', and `h' for the BSDI portion of the disk; `d' for the first other
operating system (e.g., MS/DOS), `e' for the next, and `f' for the next.
Of course, it is common to have only two operating systems present.
.PP
You should only assign to FDISK partitions those BSD partitions that you
plan to use.  You need not assign every single BSD partition.  It is very
common to see some unused partitions.
.PP
When you have set up the FDISK partition sizes and the BSD partition
assignments, use `N' to proceed to the next screeen.  You can not
proceed in certain cases (e.g., no FDISK partitions assigned).
Resolve the issue and then move on.
.br
.ne 19
.PP
The second screen resembles this:
.ft CW
.br
.nf
                           BSD Partition Sizing
FDISK
Part# Start  Length BSD Type    Length   MB  Warnings
  1       0  347165  a  Unused       0    0
                     b  Unused       0    0
                     h  Unused       0    0  BSD partitions 347165 too short
.sp 0.5
Commands:  [a-h] Set BSD partition length   [N]ext phase
           [A-H] Set BSD partition type     [P]revious page (sizing)
.sp 0.5
Command>
.sp 0.5
Directions:
   Use `a' thru `h' to set the BSD partition lengths
   Use `A' thru `H' to set the BSD partition types
   Use `N' to move to the next phase: Installing boot blocks
   Use `P' for previous page  (FDISK sizing and BSD partition assignment)
   Use `X' to abort disksetup
.ft
.sp 0.5
.br
.fi
It shows the FDISK partitions of type BSDI \- the partitions which will be
subdivided into smaller BSD partitions.  The start and length of the FDISK
partition are shown.  Only the length of the BSD partitions is shown; the
offsets are computed automatically so that the partitions are packed as
tightly as possible (and in the order shown).  Only the lengths and types
of the BSD partitions can be set on the screen.  Return the previous
screen (`P') to adjust the FDISK starting sector or length.  All start 
and length values are in terms of sectors.
.PP
Use the lower case `a' through `h' keys to set lengths; use the upper case
keys to set types.  Heed the warnings about aggregate lengths being too
long or too short.  You can not proceed to the next phase if the BSD
partitions will not fit in the FDISK partition.  Note that
.I disksetup
presents a default size value if you change the size of a BSD partition

sbin/disksetup/disksetup.cc
/*
 * Copyright (c) 1994 Berkeley Software Design, Inc. All rights reserved.
 * The Berkeley Software Design Inc. software License Agreement specifies
 * the terms and conditions for redistribution.
 */
/*	BSDI disksetup.cc,v 2.1 1995/02/03 07:14:38 polk Exp	*/

#include <err.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <ctype.h>
#include <fcntl.h>
#include "fs.h"
#include <paths.h>
#include <sys/wait.h>
#include "showhelp.h"
#include "screen.h"
#include "help.h"
#include "disk.h"
#include "field.h"
#include "util.h"

Disk disk;

#define	_PATH_BOOTANY	"/sbin/bootany"

char *path_bootany = _PATH_BOOTANY;

char *master = 0;
char *primary = 0;
char *secondary = 0;

char boot0[MAXPATHLEN];
char boot1[MAXPATHLEN];
char boot2[MAXPATHLEN];

void dumptab(char *);


O::Usage(int status)
{
    fprintf(stderr, "Usage:\n");
    fprintf(stderr,
"    %s disk                                      # Read label\n",
			prog);
    fprintf(stderr,
"    %s [-NW] disk                                # Write Disable/Enable\n",
			prog);
    fprintf(stderr,
"    %s -e [disk [xxboot bootxx [mboot]]]         # Edit disk label\n",
			prog);
    fprintf(stderr,
"    %s -i [disk]                                 # Interactive mode\n",
			prog);
    fprintf(stderr,
"    %s -R disk protofile [xxboot bootxx [mboot]] # Restore Label\n",
			prog);
    fprintf(stderr,
"    %s -B disk [ xxboot bootxx [mboot]]          # Install boot records\n",
			prog);
#if 0
    fprintf(stderr,
"    %s [-m mboot] [-b xxboot bootxx] disk        # Install boot records\n",
			prog);
#endif
    fprintf(stderr, "Additional flags:\n");
    fprintf(stderr, "    -A        Path to the bootany program\n");
    fprintf(stderr, "    -K        Only read the Kernels (internal) label\n");
    fprintf(stderr, "    -D        Only read the On Disk label\n");
    fprintf(stderr, "    -s        Only read/write incore BSD labels\n");
    fprintf(stderr, "    -n        No writes to disk (open read only and non-blocking)\n");
    fprintf(stderr, "    -I        Ignore current BSD label\n");
    fprintf(stderr, "    -I -I     Ignore current BSD label and FDISK table\n");
    fprintf(stderr, "    -F        Edit FDISK partition (implies -i)\n");
    fprintf(stderr, "    -P        Edit BSD partition (implies -i)\n");
    fprintf(stderr, "    -E        Expert mode.  Allows broken tables, etc.\n");
    fprintf(stderr, "    -M mem    Pretend physical memory is mem MB\n");
#if 0
    fprintf(stderr, "    -f        Force the installing of boot blocks (not implemented yet)\n");
    fprintf(stderr, "    -t path   Dump file system names to {path}disk\n");
    fprintf(stderr, "    -x file   Make copes of tables in file\n");
#endif

    exit(status);
}

main(int ac, char **av)
{
    int i, j;
    int e;
    char *dn;
    int c;

    char *proto = 0;
    char *output = 0;
    char *tabpath = 0;

    int mfd = -1;
    int pfd = -1;
    int sfd = -1;

    if (dn = strrchr(av[0], '/'))
	O::Program(++dn);
    else
	O::Program(av[0]);

    while ((c = getopt(ac, av, "b:efikm:np:q:st:x:A:BDEFIKM:NPRW?")) != EOF) {
	switch (c) {
	case 'A':
	    path_bootany = optarg;
	    break;
    	case 'M':
	    PhysMem(atoi(optarg) * 1024 * 1024);
	    break;
    	case 'm':	// install master boot record
	    O::Set(O::INSTALL);
	    O::Set(O::IgnoreGeometry);
    	    if (master)
		goto usage;
	    master = optarg;
	    break;
    	case 'p':	// install primary boot strap
	    O::Set(O::INSTALL);
	    O::Set(O::IgnoreGeometry);
	    O::Set(O::DontUpdateLabel);
    	    if (primary)
		goto usage;
	    primary = optarg;
	    break;
    	case 'q':	// install secondary boot strap
	    O::Set(O::INSTALL);
	    O::Set(O::IgnoreGeometry);
	    O::Set(O::DontUpdateLabel);
    	    if (secondary)
		goto usage;
	    secondary = optarg;
	    break;
    	case 'b':	// install boot straps
	    O::Set(O::INSTALL);
	    O::Set(O::IgnoreGeometry);
	    O::Set(O::DontUpdateLabel);
	    O::Set(O::DontUpdateLabel);
    	    if (ac < ++optind)
		goto usage;
    	    if (primary || secondary)
	    	goto usage;
	    primary = optarg;
	    secondary = av[optind-1];
    	    break;
	case 'e':	// Edit
	    O::Set(O::EDIT);
	    O::Set(O::IgnoreGeometry);
	    break;
	case 'F':	// Update FDISK Table
	    O::Set(O::INTERACTIVE);
	    O::Set(O::UpdateFDISK);
	    break;
	case 'P':	// Update BSD Partition Table
	    O::Set(O::INTERACTIVE);
	    O::Set(O::UpdateBSD);
	    break;
	case 'f':	// Force Boot
	    O::Set(O::ForceBoot);
	    break;
	case 's':	// Only save label in core
	    O::Set(O::InCoreOnly);
	    break;
	case 'i':	// Interactive
	    O::Set(O::INTERACTIVE);
	    break;
	case 'n':	// Nowrite
	    O::Set(O::NoWrite);
	    O::Set(O::NonBlock);
	    break;
	case 'B':	// Bootblocks
	    O::Set(O::BOOTBLKS);
	    O::Set(O::ForceBoot);
	    O::Set(O::IgnoreGeometry);
	    O::Set(O::DontUpdateLabel);
    	    if (primary || secondary || master)
	    	goto usage;
	    break;
	case 'E':	// Expert Mode
	    O::Set(O::Expert);
	    break;
	case 'I':	// Ignore label (two -> ignore fdisk label)
	    O::Set(O::Ignore);
	    O::Set(O::NonBlock);
	    break;
	case 'N':	// Make disk label not writeable
	    O::Set(O::W_DISABLE);
	    O::Set(O::NonBlock);
	    O::Set(O::Ignore);
	    O::Set(O::Ignore);
	    O::Set(O::IgnoreGeometry);
	    break;
	case 'R':	// Restore from ascii version
	    O::Set(O::RESTORE);
	    O::Set(O::NonBlock);
	    break;
	case 'W':	// Make disk label writeable
	    O::Set(O::W_ENABLE);
	    O::Set(O::NonBlock);
	    O::Set(O::Ignore);
	    O::Set(O::Ignore);
	    O::Set(O::IgnoreGeometry);
	    break;
	case 'D':	// Only read kernel label
	    O::Set(O::ReadDisk);
	    break;
	case 'K':	// Only read kernel label
	    O::Set(O::ReadKernel);
	    break;
	case 'x':	// Undocumented -- make copies of tables in savename
	    output = optarg;
	    break;
	case 't':	// Undocumented -- output tables for EZconfig
	    tabpath = optarg;
	    break;
    	usage:
	default:
	    O::Usage();
	    break;
	}
    }

    if (O::Flag(O::ReadDisk) && O::Flag(O::ReadKernel)) {
	fprintf(stderr, "Only one of -K and -D may be specified\n");
	exit(1);
    }

    if (O::Cmd() == O::UNSPEC) {
//	if (optind < ac) {
	    O::Set(O::READ);
	    O::Set(O::NoWrite);
//    	} else {
//	    O::Set(O::INTERACTIVE);
//    	}
    }

    switch (O::Cmd()) {
    case O::W_ENABLE:
    case O::W_DISABLE:
    case O::READ:
    case O::INSTALL:
	if (optind != ac - 1)
	    O::Usage();
	break;
    case O::INTERACTIVE:
	if (optind < ac - 1)
	    O::Usage();
	break;
    case O::RESTORE:
	if (optind < ac - 5 || optind > ac - 2)
	    O::Usage();
	proto = av[optind + 1];
	if (primary = av[optind + 2]) {
	    if (!(secondary = av[optind + 3]))
		O::Usage();
	    master =  av[optind + 4];
    	}
	break;
    case O::EDIT:
    case O::BOOTBLKS:
	if (optind < ac - 4 || optind > ac - 1)
	    O::Usage();
	if (primary = av[optind + 1]) {
	    if (!(secondary = av[optind + 2]))
		O::Usage();
	    master =  av[optind + 3];
    	}
	break;
    }

    if (!(dn = av[optind])) {
	if (O::Cmd() != O::INTERACTIVE && O::Cmd() != O::EDIT)
	    O::Usage(2);
	dn = ChooseDisk();
    }

    if (e = disk.Init(dn)) {
	fprintf(stderr, "initializing %s: %s\n", dn, Errors::String(e));
	exit(1);
    }

    if (O::Flag(O::UpdateFDISK) && !O::Flag(O::UpdateBSD)) {
	O::Set(O::DontUpdateLabel);
    }

    switch (O::Cmd()) {
    case O::W_ENABLE:
    	disk.WriteEnable();
	exit(0);
    case O::W_DISABLE:
    	disk.WriteDisable();
	exit(0);
    case O::READ:
    	if (!disk.label_original.Valid() || disk.label_original.Soft()) {
	    printf("# Missing disk label, default label used.\n");
	    printf("# Geometry may be incorrect.\n");
    	}
    	disk.use_fdisk = disk.has_fdisk;
	disklabel_display(disk.device, stdout, &disk.label_original, disk.PartTable());
	exit(0);
    case O::EDIT:
    	disk.use_fdisk = disk.has_fdisk;
    	edit();
    	break;
    case O::INTERACTIVE:
    	interactive();
	break;
    case O::BOOTBLKS:
    	if (!primary && disk.label_original.d_boot0) {
	    if (*disk.label_original.d_boot0 != '/')
		sprintf(boot0, "%s/%s", _PATH_BOOTSTRAPS,
				disk.label_original.d_boot0);
	    else
		strcpy(boot0, disk.label_original.d_boot0);
	    primary = boot0; 
    	}
        if (!secondary && disk.label_original.d_boot1) {
	    if (*disk.label_original.d_boot1 != '/')
		sprintf(boot1, "%s/%s", _PATH_BOOTSTRAPS,
				disk.label_original.d_boot1);
	    else
		strcpy(boot1, disk.label_original.d_boot1);
	    secondary = boot1;
        }
    	O::Set(O::INSTALL, 1);
    	disk.use_fdisk = disk.has_fdisk;
	disk.label_new = disk.label_original;
	if (!primary && !secondary)
	    getbootblocks();
    	break;
    case O::INSTALL:
    	disk.use_fdisk = disk.has_fdisk;
	disk.label_new = disk.label_original;
    	break;
    case O::RESTORE:
    	disk.use_fdisk = disk.has_fdisk;
    	break;
    default:
    	printf("That command is not implemented yet\n");
	exit(0);
    }

    if (proto) {
	FILE *fp = fopen(proto, "r");

    	if (!fp) {
	    err(1, proto);
    	}
    	if (disklabel_getasciilabel(fp, &disk.label_new, 0) == 0) {
	    errx(1, "%s: bad label prototype\n", proto);
    	}
    	fclose(fp);
    }

    BootBlock MBR;

    if (O::Flag(O::InCoreOnly) && (master || primary || secondary)) {
	print(incore_and_bootblocks_long);
	master = primary = secondary = 0;
    }

    if (master && (mfd = open(master, O_RDONLY)) < 0) {
    	if (master[0] != '/') {
	    sprintf(boot2, "%s/%s", _PATH_BOOTSTRAPS, master);
	    master = boot2;
	}
    	if ((mfd = open(master, O_RDONLY)) < 0) {
	    err(1, "MBR: %s", master);
    	}
    	if (e = MBR.Read(mfd)) {
	    errx(1, "reading MBR %s: %s\n", master, Errors::String(e));
    	}
    	close(mfd);
    } else
	MBR = disk.bootblock;

    if (primary && (pfd = open(primary, O_RDONLY)) < 0) {
    	if (primary[0] != '/') {
	    sprintf(boot0, "%s/%s", _PATH_BOOTSTRAPS, primary);
	    primary = boot0;
	}
    	if ((pfd = open(primary, O_RDONLY)) < 0) {
	    err(1, "primary bootstrap %s", primary);
    	}
    }

    if (secondary && (sfd = open(secondary, O_RDONLY)) < 0) {
    	if (secondary[0] != '/') {
	    sprintf(boot1, "%s/%s", _PATH_BOOTSTRAPS, secondary);
	    secondary = boot1;
	}
    	if ((sfd = open(secondary, O_RDONLY)) < 0) {
	    err(1, "secondary bootstrap %s", secondary);
    	}
    }

    mbpart *t;
    if (disk.use_fdisk && (t = disk.PartTable())) {
	MBR.MergePartitions(t);
    } else if (O::Flag(O::RequireBootany)) {
    	request_inform(main_bootany_no_fdisk);
    }

    //
    // If disk.bsdoff has not been set (either to where the old label
    // was, or to where we want to put the new label) we must go out
    // and calculate the value for ourselves.
    //
    if (disk.bsdoff == -1)
	disk.bsdoff = disk.LabelLocation();

    disk.label_new.ComputeSum(disk.bsdoff);

    disk.bsdoff -= LABELSECTOR * disk.secsize + LABELOFFSET;

    if (tabpath)
	dumptab(tabpath);

    if (O::Flag(O::NoWrite)) {
	if (O::Cmd() == O::INTERACTIVE)
	    disklabel_display(disk.device, stdout, &disk.label_new, disk.PartTable());
	exit(0);
    }

    if (output) {
    	disk.dfd = creat(output, 0666);
    	disk.path = output;
    	if (disk.dfd < 0) {
	    err(1, "save file %s", output);
    	}
    }

    if (master || (disk.use_fdisk && disk.PartTable())) {
    	if (O::Flag(O::RequireBootany) ||
	    memcmp(&disk.bootblock, &MBR, sizeof(BootBlock))) {
	    if (disk.has_fdisk)
		for (i = 1; i <= 4; ++i) {
#define	o	    disk.bootblock
		    if (o.Type(i) && o.Length(i)) {
			for (j = 1; j <= 4; ++j) {
			    if (MBR.Type(j) == o.Type(i) &&
				MBR.Offset(j) == o.Offset(i) &&
				MBR.Length(j) == o.Length(i))
				    break;
			}
			if (j > 4)
			    printf("Current FDISK partition %d of type %s"
				   "(0x%02x) will be destroyed.\n",
				    i, PTypeFind(o.Type(i)).name, o.Type(i));
		    }
#undef	o
		}
	    if (request_yesno(okay_to_write_mbr, 1)) {
		if (lseek(disk.dfd, 0, L_SET) == -1) {
		    warn("%s: seeking to MBR (not written)", disk.path);
		} else {
		    disk.WriteEnable();
		    MBR.Write(disk.dfd);
		    fsync(disk.dfd);
		    if (O::Flag(O::RequireBootany)) {
			char cmdbuf[1024];
			sprintf(cmdbuf, "%s -n -i %s", path_bootany, disk.path);
			system(cmdbuf);
		    }
		    fsync(disk.dfd);
		    disk.WriteDisable();
		}
	    }
    	} else {
	    request_inform(mbr_stays_same);
    	}
    }

    if (O::Flag(O::DontUpdateLabel) && sfd < 0 && pfd < 0) {
	//
	// If we are not going to update the label, the primary,
	// or the secondary bootstraps, might as well exit.
	//
	exit(0);
    }

    unsigned char bootarea[BBSIZE];

    memset(bootarea, 0, sizeof(bootarea));

    //
    // We start assuming we will only write the disk label
    // We increase the size as we figure out more to write.
    //
    int start = LABELSECTOR * disk.secsize + LABELOFFSET;
    int stop = LABELSECTOR * disk.secsize + LABELOFFSET + sizeof(DiskLabel);

    if (sfd >= 0) {
	if (read(sfd, &bootarea[disk.secsize], BBSIZE - disk.secsize) < 0) {
    	    fprintf(stderr, "%s: %s while reading\n", secondary, strerror(errno));
	    exit(1);
    	}
    	stop = BBSIZE;
	close(sfd);
    }

    if (pfd >= 0) {
	int o = (disk.bsdoff == 0 && disk.use_fdisk) ? 15 * disk.secsize : 0;
    	if (o == 0) {
	    start = 0;
    	} else if (sfd < 0) {
	    print(need_secondary_bootstrap);
	    exit(1);
    	}
	if (read(pfd, &bootarea[o], disk.secsize) < 0) {
    	    fprintf(stderr, "%s: %s while reading\n", primary, strerror(errno));
	    exit(1);
    	}
	close(pfd);
    }

    //
    // Okay to check space even if not installing label
    //
    unsigned char *p = bootarea + LABELSECTOR * disk.secsize + LABELOFFSET;
    DiskLabel *dl = (DiskLabel *)p;
    while (p < (unsigned char *)(dl+1)) {
	if (*p++) {
	    print(no_room_for_label);
	    exit(1);
    	}
    }

    if (O::Flag(O::DontUpdateLabel) && !output && !proto
				    && (sfd >= 0 || pfd >= 0)) {
	//
	// Okay, we are not supposed to modify the label.  We just read
	// the sector where the label would go and read it up.
	//
    	DiskLabel label;
    	off_t offset;

    	//
    	// This could be a problem on an unlabled disk...
    	//
	int e = label.Read(disk.dfd, disk.bsdoff + LABELSECTOR * disk.secsize
					         + LABELOFFSET, &offset);
    	if (e) {
	    fprintf(stderr, "reading existing disk label from %s: %s\n", disk.path, Errors::String(e));
	    exit(1);
    	}
    	if (offset != disk.bsdoff + LABELSECTOR * disk.secsize + LABELOFFSET) {
    	    request_inform(use_old_label);
    	}
    	label.ComputeSum(disk.bsdoff + LABELSECTOR*disk.secsize + LABELOFFSET);
    	*dl = label;
    } else if (!O::Flag(O::DontUpdateLabel)) {
	*dl = disk.label_new;
    }

    printf("\n");

    if (disk.label_original.Valid() && !disk.label_original.Soft()) {
	for (i = 0; i < 8; ++i) {
#define	o   disk.label_original.d_partitions[i]
	    if (o.p_size && o.p_fstype != FS_SWAP && o.p_fstype != FS_UNUSED) {
		for (j = 0; j < 8; ++j) {
#define	n	dl->d_partitions[j]
		    if (o.p_size == n.p_size &&
			o.p_offset == n.p_offset &&
			o.p_fstype == n.p_fstype)
			    break;
		}
		if (j >= 8)
		    printf("Current BSD partition %c will be destroyed.\n",
			    i + 'a');
	    } 
#undef n
#undef o
	}
    }


    if (O::Flag(O::InCoreOnly) ||
	request_yesno((sfd < 0 && pfd < 0) ? okay_to_write_dl
					   : okay_to_write_dlbb, 1)) {

	//
	// First write the label in core
	//
    	if (dl->WriteInternal(disk.dfd)) {
	    warn("%s: writing internal label", disk.path);
	}

    	if (!O::Flag(O::InCoreOnly)) {
	    disk.WriteEnable();

    	    int o;
    	    char tbuf[disk.secsize];

    	    if (o = (start % disk.secsize)) {
	    	if (lseek(disk.dfd, disk.bsdoff + start - o, L_SET) == -1 ||
		    read(disk.dfd, tbuf, disk.secsize) != disk.secsize) {

		    warn("%s: merging boot blocks", disk.path);
		    fsync(disk.dfd);
		    exit(0);
    	    	}
		memcpy(bootarea + start - o, tbuf, o);
    	    	start -= o;
    	    }

    	    if (o = (stop % disk.secsize)) {
	    	if (lseek(disk.dfd, disk.bsdoff + stop - o, L_SET) == -1 ||
		    read(disk.dfd, tbuf, disk.secsize) != disk.secsize) {

		    warn("%s: merging boot blocks", disk.path);
		    fsync(disk.dfd);
		    exit(0);
    	    	}
		memcpy(bootarea + stop, tbuf + o, disk.secsize - o);
    	    	stop += disk.secsize - o;
    	    }

	    if (lseek(disk.dfd, disk.bsdoff + start, L_SET) == -1 ||
		write(disk.dfd, bootarea + start, stop-start) != stop-start) {
		warn("%s: writing boot blocks", disk.path);
	    }
	    disk.WriteDisable();
    	}
	fsync(disk.dfd);
    }
    exit(0);
}

void
dumptab(char *tabpath)
{
    FILE *fp;
    int i;
    char tabname[strlen(tabpath) + strlen(disk.device) + 1];
    
    strcpy(tabname, tabpath);
    strcat(tabname, disk.device);

    if ((fp = fopen(tabname, "w")) == NULL)
	return;
    for (i = 0; i < 8 && disk.bsd[i].number; ++i) {
	FileSystem &f = disk.bsd[i];
	if (f.number == 3)
	    continue;
	if (f.mount[0] == '\0')
	    strcat(f.mount, "UNKNOWN");
	fprintf(fp, "%s%c %s %s\n", disk.device, f.number + 'a' - 1, f.mount,
				    FTypeFind(f.type).name);
    }
    fclose(fp);
}
